tiled/storage.py

```diff
@@ -34,40 +42,96 @@ def path(self):
         return path_from_uri(self.uri)
 
 
-@dataclasses.dataclass(frozen=True)
-class EmbeddedSQLStorage(Storage):
-    "File-based SQL database storage location"
+def _ensure_writable_location(uri: str) -> Path:
+    "Ensure path is writable to avoid a confusing error message from driver."
+    filepath = path_from_uri(uri)
+    directory = Path(filepath).parent
+    if directory.exists():
+        if not os.access(directory, os.X_OK | os.W_OK):
+            raise ValueError(
+                f"The directory {directory} exists but is not writable and executable."
+            )
+        if Path(filepath).is_file() and (not os.access(filepath, os.W_OK)):
+            raise ValueError(f"The path {filepath} exists but is not writable.")
+    else:
+        raise ValueError(f"The directory {directory} does not exist.")
+    return filepath
 
 
 @dataclasses.dataclass(frozen=True)
 class SQLStorage(Storage):
+    "General purpose SQL database storage with connection pooling"
+
+    @abstractmethod
+    def create_adbc_connection(self) -> "adbc_driver_manager.dbapi.Connection":
+        "Create a connection to the database."
+
+        raise NotImplementedError("Subclasses must implement this method.")
+
+    @property
+    def dialect(self) -> str:
+        "The database dialect (e.g. 'postgresql', 'sqlite', or 'duckdb')."
+        return urlparse(self.uri).scheme
+
+    @functools.cached_property
+    def _adbc_connection(self) -> "adbc_driver_manager.dbapi.Connection":
+        "A persistent connection to the database."
+        return self.create_adbc_connection()
+
+    @functools.cached_property
+    def _connection_pool(self) -> "sqlalchemy.pool.QueuePool":
+        creator = self._adbc_connection.adbc_clone
+        if (self.dialect == "duckdb") or (":memory:" in self.uri):
+            return sqlalchemy.pool.StaticPool(creator)
+        else:
+            return sqlalchemy.pool.QueuePool(creator, pool_size=3, max_overflow=0)
+
+    def connect(self) -> "adbc_driver_manager.dbapi.Connection":
+        "Get a connection from the pool."
+        return self._connection_pool.connect()
+
+    def dispose(self) -> None:
+        "Close all connections and dispose of the connection pool."
+        self._connection_pool.dispose()
+        self._adbc_connection.close()
+
+
+@dataclasses.dataclass(frozen=True)
+class EmbeddedSQLStorage(SQLStorage):
     "File-based SQL database storage location"
+
+    def create_adbc_connection(self) -> "adbc_driver_manager.dbapi.Connection":
+        filepath = _ensure_writable_location(self.uri)
+
+        if self.uri.startswith("duckdb:"):
+            import adbc_driver_duckdb.dbapi
+
+            return adbc_driver_duckdb.dbapi.connect(str(filepath))
+
+        elif self.uri.startswith("sqlite:"):
+            import adbc_driver_sqlite.dbapi
+
+            return adbc_driver_sqlite.dbapi.connect(str(filepath))
```

Comment:
Some kind of check here, such as `else: raise ValueError(...)` would be appropriate since the object and method are public and might receive a different scheme.